### PR TITLE
Make a 4.0 release

### PR DESCRIPTION
**Pull request by vishnugopal**
**Short description:**

This is a release PR to make a 4.0 release to change the way files are imported in Twist. A few types have been added here to make Twist error-free.

This is still WIP!

**PR Checklist:**

-   [x] Added tests for bugs / new features
-   [x] Updated docs (storybooks, readme)
-   [x] Described changes in CHANGELOG.md
-   [x] Executed `npm run check` and made sure no errors / warnings were shown
-   [x] Executed `npm run test` and made sure all tests are passing
-   [x] Bumped version in package.json
-   [x] Updated all static build artifacts (`npm run build-all`)

### CODE REVIEW

**Comment by vishnugopal** [comment]
@henningmu The entire lib/ directory and the webpack workflow can technically be deleted since Twist will now import from dist/, but I'd like to hold on to it for some time so that if there's any regression in Twist, we can roll back faster. It's easy to remove later.  You can verify this by manually linking the Twist PR and deleting the lib/ directory.

**Comment by henningmu** [comment]
Works for me 👍

btw: a release can be done by creating a release on GH (or pushing a tag). We have a GH action that then takes care of publishing 👍

**Comment by proxi** [comment]
> A few types have been added here to make Twist error-free.

@vishnugopal @henningmu  I'm a bit confused as to what happened here, as we already did have all the types (specified as JSDoc, with noImplicitAny enabled). Your conversion to tsc seems to have remove most of them, a big step backwards?

Also, keeping type-only JSDoc comments after conversion seems confusing.

**Comment by proxi** [comment]
As for 4.0 release plan - I would not push a major version without fixing strict mode, as this will change public API in several places.

**Comment by henningmu** [comment]
> I'm a bit confused as to what happened here

Shoot that's my fault. I mainly tested that nothing breaks but didn't review #196 thoroughly enough and didn't notice the degradation in typings there.

@vishnugopal can you fix this please before doing a release of the code?

> As for 4.0 release plan

Usage of reactist is quite low. To be honest I wouldn't mind to bump 3 majors in a week if necessary. Whatever works better. If we can batch stuff together that's appreciated as well 👍

**Comment by vishnugopal** [comment]
Hi! I staggered this release in this fashion because I wanted a version where I could move to tsdx and test integration with Twist as soon as possible, i.e. release a 4.0 where Twist could migrate, and then slowly work on bringing the strict types back, but I can do the other way around too by working on strictness first as @proxi suggested.

Let me rework this to make `strict: true` in reactist then!

**Comment by vishnugopal** [comment]
Just a note that I'm making progress here, but it's slow. Around 235 type errors to be gone now (I started from around 400+) :-). Think it'll take me 2-3 days more to get `strict: true` working!

I'm also finding some null & undefined errors and fixing them, so that might also increase the testing scope a bit as this now has a few code changes too.

**Comment by vishnugopal** [comment]
@henningmu suggested that I split the `strict: true` work into smaller PRs as reviewing all of em together might be a mess. So I've raised a first PR here: https://github.com/Doist/reactist/pull/199 

The plan is to have a number of such PRs targeting this branch until we can finally enable strict: true here. 

Hope that sounds ok @proxi and @gnapse

**Comment by proxi** [comment]
> release a 4.0 where Twist could migrate, and then slowly work on bringing the strict types back

We cannot migrate without types, Twist-Web already requires types (noImplicitAny and soon typescript/no-explicit-any).

> Just a note that I'm making progress here, but it's slow. Around 235 type errors to be gone now (I started from around 400+) :-).

@vishnugopal 3.1 release was not strict, so it's ok to start without it. But it was `noImplicitAny` and it should stay this way. I'm a bit puzzled as to what you are working on: all the types are already there in JSDoc comments, you should just change syntax, e.g:

```
before:
/**
     * @param {HTMLLIElement} tooltip
     */
    _updateTooltipRef = (tooltip) => {
        ;(this as any).tooltip = tooltip
    }

after:
    _updateTooltipRef = (tooltip: HTMLLIElement) => {
        ;(this as any).tooltip = tooltip
    }
```

There's no need to keep type-JSDoc in .tsx files and they should be removed after conversion.

> Think it'll take me 2-3 days more to get strict: true working!

If you keep the types we already have, there's only about 18 strict-errors left. There was some progress against 3.1 branch [here](https://github.com/Doist/reactist/pull/193).

To sum up, you should:
- convert JSDoc comments to TypeScript type syntax
- enable noImplicitAny (if you convert JSDoc correctly there should be 0 errors)
- do `strict: true` as a separate PR later (there are about 18 errors to fix)

**Comment by vishnugopal** [comment]
> To sum up, you should:

> * convert JSDoc comments to TypeScript type syntax
> * enable noImplicitAny (if you convert JSDoc correctly there should be 0 errors)
> * do strict: true as a separate PR later (there are about 18 errors to fix)

I also upgraded eslint and prettier to support new typescript syntax, and there were a few niggles as well there. I've reduced error count in total to:

> Found 169 errors.

once [this branch is merged](https://github.com/Doist/reactist/pull/199), which is good progress for me 🙂

> If you keep the types we already have, there's only about 18 strict-errors left. There was some progress against 3.1 branch here.

I'll look into that branch too, thanks!

>  But it was noImplicitAny and it should stay this way. I'm a bit puzzled as to what you are working on: all the types are already there in JSDoc comments, you should just change syntax, e.g:

I think most of my time until now went in converting the files to tsdx, making it work with the new rollup builder, and then the manual process of converting the files one by one. There were a bunch of files that were not typed (all the test files for e.g.) and that took some doing too. I did search for some automatic way to convert these to Typescript but couldn't find any, so for now it's a manual process for me!

**Comment by proxi** [comment]
I wouldn't worry about test files too much, not worth spending time there. Also not doable until main source is types.

> I've reduced error count in total to: Found 169 errors.

To be honest I don't think you've reduced it, since it was zero with types and now it's 169 without types ;) But https://github.com/Doist/reactist/pull/199 is starting to look better

It's very hard for me to review all the changes here, since there are so many and dev is broken. I will re-review everything once the types are back.

File: src/components/__tests__/Tooltip.test.tsx [hunk 1]
```diff
@@ -188,22 +201,31 @@ describe('Tooltip', () => {
             const currentProps = tooltip.props
             const shouldUpdate = tooltip.shouldComponentUpdate(
                 currentProps,
-                currentState
+                currentState,
+                undefined
```
Comment (proxi) [inline src/components/__tests__/Tooltip.test.tsx:205]: What and why is this third argument? shouldComponentUpdate only takes two (nextProps, nextState).

Comment (vishnugopal) [reply src/components/__tests__/Tooltip.test.tsx:205]: The types somehow seem to say it accepts 3:

```
shouldComponentUpdate?(nextProps: Readonly<P>, nextState: Readonly<S>, nextContext: any): boolean;
```

Comment (proxi) [reply src/components/__tests__/Tooltip.test.tsx:205]: Something's clearly off because we only have 2 in Tooltip.tsx. But I guess we can live with this.

Alternative could be `shallow(getTooltip()).instance() as Tooltip` but I'm not sure it's better.

**Comment by vishnugopal** [comment]
Just a note on today's update: [PR #200](https://github.com/Doist/reactist/pull/200) is up for review.

The number of errors is now down to:

> Found 41 errors.

Yay! If all goes well, I should be able to wrap this up tomorrow & get this release-ready.

**Comment by vishnugopal** [comment]
This should now be ready for a final check with Twist. I've removed the webpack builder from package.json, but it's still available in the repository in case something happens after merge and we'd like to roll back to the older imports.
